README
autocutsel
http://www.nongnu.org/autocutsel/
https://github.com/sigmike/autocutsel
Copyright (c) 2001-2021 - Michael Witrant

X servers use two schemes to copy text between applications. The first one
(old and deprecated) is the cutbuffer. It is a simple buffer in which any
application can store text. The other scheme is the selection and works
differently. There may be many selections in a single server. An
application does not copy data in a selection, it "owns" it. When
another application wants to retrieve the content of a selection, it
asks the owner.

Recent desktop applications (GNOME, KDE, ...) use two selections: the
PRIMARY and the CLIPBOARD. The PRIMARY selection is used when you select
some text with the mouse. You usually paste it using the middle button. The
CLIPBOARD selection is used when you copy text by using, for example,
the Edit/Copy menu. You may paste it using the Edit/Paste menu.

Windows VNC clients keep the Windows clipboard synchronized with the
cutbuffer, but not with the selections. And since recent applications
don't use the cutbuffer, the server's CLIPBOARD is never synchronized
with Windows' one.

Autocutsel tracks changes in the server's cutbuffer and CLIPBOARD
selection. When the CLIPBOARD is changed, it updates the cutbuffer. When
the cutbuffer is changed, it owns the CLIPBOARD selection. The cutbuffer
and CLIPBOARD selection are always synchronized. Since the VNC client
synchronizes the Windows' clipboard and the server's cutbuffer, all
three "clipboards" are always kept synchronized. When you copy some text in
Windows, the cutbuffer and the CLIPBOARD selection are updated. When you
copy text on the server using either the cutbuffer or the CLIPBOARD
selection, the Windows's clipboard is always updated.

To build autocutsel:
 ./configure
 make

If you downloaded autocutsel from git you first have to run ./bootstrap to
build the configure script.

If configure fails because it can't find X or Xaw libraries, you may have to
install the devel packages of these libraries.

On Ubuntu, you may need to install these packages :
  sudo apt-get install libc6-dev libxtst-dev xorg-dev

To install it (as root):
 make install

To run it:
 autocutsel

By default, autocutsel synchronizes the CLIPBOARD selection. You may want to use
the PRIMARY selection instead, if your applications use Motif or another toolkit
not using the CLIPBOARD. To do that, simply run
 autocutsel -selection PRIMARY

Some softwares (like Open Office Writer) have trouble when the PRIMARY selection
is requested before the mouse button is released. As a workaround, you can run
autocutsel with the "-buttonup" option and it will only get the selection when
the first mouse button is not pressed.

Any feedback is welcome.

Michael Witrant <mike at lepton.fr>


    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


ChangeLog
v0.10.1: 2021-03-03
  - fixed allocation error (Jeremie Courreges-Anglas)
	- updated configure scripts

v0.10.0: 2014-05-04
  - set WM_NAME, _NET_WM_NAME, and _NET_WM_PID (Michael Stapelberg)
  - support for UTF-8 strings
  - minor fixes

v0.9.1: 2013-11-19
  - fixed cutsel segfault for empty cut buffer (Jurij Smakov)
  - built package with libtool 2.4.2.418 to support ppc64le

v0.9.0: 2006-11-05
  - really works on 64 bits linux
  - added test suite
  - added workaround for open office writer (Bob Kinney)
  - fixed compilation problem on NetBSD (Jeremy C. Reed)

v0.8.1:
  - fixed a bug on 64 bits systems
  - improved debug messages

v0.8:
  - removed blank square
  - added option -fork
  - removed dmalloc support

v0.7.1: 2004-11-03
  - fixed compilation problems on some platforms
  - ported 0.7.0 changes to cutsel.c

v0.7.0: 2004-09-11
  - fixed a bug when the first copy comes from the cutbuffer (jluu)
  - the -selection parameter now works
  - changed default selection to CLIPBOARD instead of PRIMARY (jluu)
    It means you now have to *COPY* and *PASTE* texts to have them handled
    by autocutsel. Just selecting text will only work in a few applications.
    If you want the old behavior, run "autocusel -selection PRIMARY"

v0.6.2: 2002-06-03
  - added a cutsel tools to change/retrieve the cutbuffer
    or the selection
  - fixed minor bugs

v0.6.1: 2002-06-02
  - uses the length of values instead of '\0' at the end

v0.6: 2002-05-27
  - fixed a memory leak

v0.5: 2002-05-26
  - fixed a bug the prevented selection to be owned when
    the cutbuffer was modified
  - added dmalloc support (to find memory leaks)

v0.4: 2002-05-19
  - perform a last check before owning the selection
    to handle properly applications updating both cutbuffer
    and selection
  - autoconf & automake scripts
  - improved debug messages
  - memory leak fixed?
  - minor bugfixes

v0.3:
  - applied a patch from Mark Warburton to ignore the cutbuffer
    value if it's empty
  - added -verbose (displays what -debug displayed before)
  - changed -debug to display more debugging informations
  - modified the synchronization process. it should fix some
    update problems.

v0.2: 2001-07-15
  - command line options: cutbuffer, selection, pause and debug
  - removed bell on error. Seems to help with VNC connections.

v0.1: 2001-07-11
  - initial release


autocutsel.c
/*
 * autocutsel by Michael Witrant <mike @ lepton . fr>
 * Synchronizes the cutbuffer and the selection
 * Copyright (c) 2001-2021 Michael Witrant.
 *
 * Most code taken from:
 * * clear-cut-buffers.c by "E. Jay Berkenbilt" <ejb @ ql . org>
 *   in this messages:
 *     http://boudicca.tux.org/mhonarc/ma-linux/2001-Feb/msg00824.html
 *
 * * xcutsel.c by Ralph Swick, DEC/Project Athena
 *   from the XFree86 project: http://www.xfree86.org/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * This program is distributed under the terms
 * of the GNU General Public License (read the COPYING file)
 *
 */

#include "common.h"

static XrmOptionDescRec optionDesc[] = {
  {"-selection", "selection", XrmoptionSepArg, NULL},
  {"-select",    "selection", XrmoptionSepArg, NULL},
  {"-sel",       "selection", XrmoptionSepArg, NULL},
  {"-s",         "selection", XrmoptionSepArg, NULL},
  {"-cutbuffer", "cutBuffer", XrmoptionSepArg, NULL},
  {"-cut",       "cutBuffer", XrmoptionSepArg, NULL},
  {"-c",         "cutBuffer", XrmoptionSepArg, NULL},
  {"-debug",     "debug",     XrmoptionNoArg,  "on"},
  {"-d",         "debug",     XrmoptionNoArg,  "on"},
  {"-verbose",   "verbose",   XrmoptionNoArg,  "on"},
  {"-v",         "verbose",   XrmoptionNoArg,  "on"},
  {"-fork",      "fork",      XrmoptionNoArg,  "on"},
  {"-f",         "fork",      XrmoptionNoArg,  "on"},
  {"-pause",     "pause",     XrmoptionSepArg, NULL},
  {"-p",         "pause",     XrmoptionSepArg, NULL},
  {"-buttonup",  "buttonup",  XrmoptionNoArg,  "on"},
};

int Syntax(char *call)
{
  fprintf (stderr,
    "usage:  %s [-selection <name>] [-cutbuffer <number>]"
    " [-pause <milliseconds>] [-debug] [-verbose] [-fork] [-buttonup]\n",
    call);
  exit (1);
}

#define Offset(field) XtOffsetOf(OptionsRec, field)

static XtResource resources[] = {
  {"selection", "Selection", XtRString, sizeof(String),
    Offset(selection_name), XtRString, "CLIPBOARD"},
  {"cutBuffer", "CutBuffer", XtRInt, sizeof(int),
    Offset(buffer), XtRImmediate, (XtPointer)0},
  {"debug", "Debug", XtRString, sizeof(String),
    Offset(debug_option), XtRString, "off"},
  {"verbose", "Verbose", XtRString, sizeof(String),
    Offset(verbose_option), XtRString, "off"},
  {"fork", "Fork", XtRString, sizeof(String),
    Offset(fork_option), XtRString, "off"},
  {"kill", "kill", XtRString, sizeof(String),
    Offset(kill), XtRString, "off"},
  {"pause", "Pause", XtRInt, sizeof(int),
    Offset(pause), XtRImmediate, (XtPointer)500},
  {"buttonup", "ButtonUp", XtRString, sizeof(String),
    Offset(buttonup_option), XtRString, "off"},
};

#undef Offset

static void CloseStdFds()
{
  int fd;

  for (fd = 0; fd < 3; fd++)
    close (fd);
}

static void Terminate(int caught)
{
  exit (0);
}

static void TrapSignals()
{
  int catch;
  struct sigaction action;

  sigemptyset (&action.sa_mask);
  action.sa_flags = 0;
  for (catch = 1; catch < NSIG; catch++) {
    if (catch != SIGKILL) {
      action.sa_handler = Terminate;
      sigaction (catch, &action, (struct sigaction *) 0);
    }
  }
}

// Called when we no longer own the selection
static void LoseSelection(Widget w, Atom *selection)
{
  if (options.debug)
    printf("Selection lost\n");
  options.own_selection = 0;
}

// Returns true if value (or length) is different
// than current ones.
static int ValueDiffers(char *value, int length)
{
  return (!options.value ||
    length != options.length ||
    memcmp(options.value, value, options.length));
}

// Update the current value
static void ChangeValue(char *value, int length)
{
  if (options.value)
    XtFree(options.value);

  options.length = length;
  options.value = XtMalloc(options.length);
  if (!options.value)
    printf("WARNING: Unable to allocate memory to store the new value\n");
  else {
    memcpy(options.value, value, options.length);

    if (options.debug) {
      printf("New value saved: ");
      PrintValue(options.value, options.length);
      printf("\n");
    }
  }
}

// Called just before owning the selection, to ensure we don't
// do it if the selection already has the same value
static void OwnSelectionIfDiffers(Widget w, XtPointer client_data,
                                  Atom *selection, Atom *type, XtPointer value,
                                  unsigned long *received_length, int *format)
{
  int length = *received_length;

  if (*type == 0 ||
      *type == XT_CONVERT_FAIL ||
      length == 0 ||
      ValueDiffers(value, length)) {
    if (options.debug)
      printf("Selection is out of date. Owning it\n");

    if (options.verbose)
    {
      printf("cut -> sel: ");
      PrintValue(options.value, options.length);
      printf("\n");
    }

    if (XtOwnSelection(box, options.selection,
        0, //XtLastTimestampProcessed(dpy),
        ConvertSelection, LoseSelection, NULL) == True) {
      if (options.debug)
        printf("Selection owned\n");
      options.own_selection = 1;
    }
    else
      printf("WARNING: Unable to own selection!\n");
  }
  XtFree(value);
}

// Look for change in the buffer, and update the selection if necessary
static void CheckBuffer()
{
  char *value;
  int length;

  value = XFetchBuffer(dpy, &length, buffer);

  if (length > 0 && ValueDiffers(value, length)) {
    if (options.debug) {
      printf("Buffer changed: ");
      PrintValue(value, length);
      printf("\n");
    }

    ChangeValue(value, length);
    XtGetSelectionValue(box, selection, XInternAtom(dpy, "UTF8_STRING", False),
      OwnSelectionIfDiffers, NULL,
      CurrentTime);
  }

  XFree(value);
}

// Called when the requested selection value is available
static void SelectionReceived(Widget w, XtPointer client_data, Atom *selection,
                              Atom *type, XtPointer value,
                              unsigned long *received_length, int *format)
{
  int length = *received_length;

  if (*type != 0 && *type != XT_CONVERT_FAIL) {
    if (length > 0 && ValueDiffers(value, length)) {
      if (options.debug) {
        printf("Selection changed: ");
        PrintValue((char*)value, length);
        printf("\n");
      }

      ChangeValue((char*)value, length);
      if (options.verbose) {
        printf("sel -> cut: ");
        PrintValue(options.value, options.length);
        printf("\n");
      }
      if (options.debug)
        printf("Updating buffer\n");

      XStoreBuffer(XtDisplay(w),
             (char*)options.value,
             (int)(options.length),
             buffer );

      XtFree(value);
      return;
    }
  }
  XtFree(value);

  // Unless a new selection value is found, check the buffer value
  CheckBuffer();
}

// Called each <pause arg=500> milliseconds
void timeout(XtPointer p, XtIntervalId* i)
{
  if (options.own_selection)
    CheckBuffer();
  else {
    int get_value = 1;
    if (options.buttonup) {
      int screen_num = DefaultScreen(dpy);
      int root_x, root_y, win_x, win_y;
      unsigned int mask;
      Window root_wnd, child_wnd;
      XQueryPointer(dpy, RootWindow(dpy,screen_num), &root_wnd, &child_wnd,
        &root_x, &root_y, &win_x, &win_y, &mask);
      if (mask & (ShiftMask | Button1Mask))
        get_value = 0;
    }
    if (get_value)
      XtGetSelectionValue(box, selection, XInternAtom(dpy, "UTF8_STRING", False),
        SelectionReceived, NULL,
        CurrentTime);
  }

  XtAppAddTimeOut(context, options.pause, timeout, 0);
}

int main(int argc, char* argv[])
{
  Widget top;
  top = XtVaAppInitialize(&context, "AutoCutSel",
        optionDesc, XtNumber(optionDesc), &argc, argv, NULL,
        XtNoverrideRedirect, True,
        XtNgeometry, "-10-10",
        NULL);

  if (argc != 1) Syntax(argv[0]);

  XtGetApplicationResources(top, (XtPointer)&options,
          resources, XtNumber(resources),
          NULL, ZERO );


  if (strcmp(options.debug_option, "on") == 0)
    options.debug = 1;
  else
    options.debug = 0;

  if (strcmp(options.verbose_option, "on") == 0)
    options.verbose = 1;
  else
    options.verbose = 0;

  if (options.debug || options.verbose)
    printf("autocutsel v%s\n", VERSION);

  if (strcmp(options.buttonup_option, "on") == 0)
    options.buttonup = 1;
  else
    options.buttonup = 0;

  if (strcmp(options.fork_option, "on") == 0) {
    options.fork = 1;
    options.verbose = 0;
    options.debug = 0;
  }
  else
    options.fork = 0;

  if (options.fork) {
    if (getppid () != 1) {
#ifdef SETPGRP_VOID
      setpgrp();
#else
      setpgrp(0, 0);
#endif
      switch (fork()) {
      case -1:
        fprintf (stderr, "could not fork, exiting\n");
        return errno;
      case 0:
        sleep(3); /* Wait for father to exit */
        chdir("/");
        TrapSignals();
        CloseStdFds();
        break;
      default:
        return 0;
      }
    }
  }

  options.value = NULL;
  options.length = 0;

  options.own_selection = 0;

  box = XtCreateManagedWidget("box", boxWidgetClass, top, NULL, 0);
  dpy = XtDisplay(top);

  selection = XInternAtom(dpy, options.selection_name, 0);
  options.selection = selection;
  buffer = 0;

  options.value = XFetchBuffer(dpy, &options.length, buffer);

  XtAppAddTimeOut(context, options.pause, timeout, 0);
  XtRealizeWidget(top);
  XUnmapWindow(XtDisplay(top), XtWindow(top));

  /* Set up window properties so that the PID of the relevant autocutsel
   * process is known and autocutsel windows can be identified as such when
   * debugging. */
  const int _NET_WM_NAME = 0;
  const int UTF8_STRING = 1;
  const int WM_NAME = 2;
  const int STRING = 3;
  const int _NET_WM_PID = 4;
  const int CARDINAL = 5;

  Atom atoms[6];
  char *names[6] = {
      "_NET_WM_NAME",
      "UTF8_STRING",
      "WM_NAME",
      "STRING",
      "_NET_WM_PID",
      "CARDINAL",
  };
  XInternAtoms(dpy, names, 6, 0, atoms);

  const char *window_name = "autocutsel";
  pid_t pid = getpid();

  XChangeProperty(dpy, XtWindow(top), atoms[_NET_WM_NAME], atoms[UTF8_STRING], 8, PropModeReplace, (const unsigned char*)window_name, strlen(window_name));
  XChangeProperty(dpy, XtWindow(top), atoms[WM_NAME], atoms[STRING], 8, PropModeReplace, (const unsigned char*)window_name, strlen(window_name));
  XChangeProperty(dpy, XtWindow(top), atoms[_NET_WM_PID], atoms[CARDINAL], 32, PropModeReplace, (const unsigned char*)&pid, 1);

  XtAppMainLoop(context);

  return 0;
}
